shared-internals/src/index.js

```diff
@@ -0,0 +1,109 @@
+/**
+ * Get the direct parent of a `VNode`.
+ * @param {import('../../src/internal').VNode} vnode
+ * @returns {import('../../src/internal').VNode | null}
+ */
+export function getParent(vnode) {
+	return vnode._parent;
+}
+
+/**
+ * Return the `Component` instance associated with a rendered `VNode`.
+ * @param {import('../../src/internal').VNode} vnode
+ * @returns {import('../../src/internal').Component | null}
+ */
+export function getComponent(vnode) {
+	return vnode._component;
+}
+
+/**
+ * Return the rendered DOM node associated with a rendered `VNode`.
+ *
+ * "Associated" here means either the DOM node directly output as a result of
+ * rendering the vnode (for DOM vnodes) or the first DOM node output by a
+ * child vnode for component vnodes.
+ *
+ * @param {import('../../src/internal').VNode} vnode
+ * @returns {import('../../src/internal').PreactElement | Text | null}
+ */
+export function getDom(vnode) {
+	return vnode._dom;
+}
+
+/**
+ * Return the child `VNodes` associated with a rendered `VNode`.
+ * @param {import('../../src/internal').VNode} vnode
+ * @returns {Array<import('../../src/internal').VNode | null | undefined>}
+ */
+export function getChildren(vnode) {
+	return vnode._children;
+}
+
+/**
+ * Return the `VNode` of a `component` when it was last rendered.
+ * @param {import('../../src/internal').Component} component
+ * @returns {import('../../src/internal').VNode}
+ */
+export function getComponentVNode(component) {
+	return component._vnode;
+}
+
+// Options getters/setters
+
+/**
+ * Return private `options._diff` hook that's called before `diff()`.
+ * @param {import('../../src/internal').Options} options
+ * @returns {import('../../src/internal').Options["_diff"]}
+ */
+export function getOptionsDiff(options) {
+	return options._diff;
+}
+
+/**
+ * Return private `options._commit` hook that's called at the end of
+ * each commit.
+ * @param {import('../../src/internal').Options} options
+ * @returns {import('../../src/internal').Options["_commit"]}
+ */
+export function getOptionsCommit(options) {
+	return options._commit;
+}
+
+/**
+ * Return private `options._root` hook that's called when a tree
+ * is rendered from the top via `render(vnode, dom)`.
+ * @param {import('../../src/internal').Options} options
+ * @returns {import('../../src/internal').Options["_root"]}
+ */
+export function getOptionsRoot(options) {
+	return options._root;
+}
+
+/**
+ * Set private `options._diff` hook that's called before `diff()`.
+ * @param {import('../../src/internal').Options} options
+ * @param {import('../../src/internal').Options["_diff"]} fn
+ */
+export function setOptionsDiff(options, fn) {
+	options._diff = fn;
+}
+
+/**
+ * Return private `options._commit` hook that's called at the end of
+ * each commit.
+ * @param {import('../../src/internal').Options} options
+ * @param {import('../../src/internal').Options["_commit"]} fn
+ */
+export function setOptionsCommit(options, fn) {
```

Comment:
Hmm is little nervous exposing this option before we do 2 phase render. By exposing the option we also expose it's contract (i.e. arguments) and I suspect the commit option will change after 2 phase (personally think it might be interesting to rewrite commit option to enable different renders).

Which libraries is this need for? Devtools? I also really want the devtools lol

Reply:
That's a very good point! For now the devtools "bridge" will likely live in `preact/debug` so there is no immediate need for the devtools expose this. We do have one known outside consumer though: https://github.com/duxiaofeng-github/preact-responder-event-plugin/blob/master/src/index.ts#L8 . Same is true for the Composition API in #1923 .

I'm wondering if a 2-phase renderer would be released as semver major or minor. It shouldn't be a visible breaking change for users from an API point of view though.

I'm admittedly a bit torn. On the one hand I'm ok with doing breaking changes here for our shared internals, but on the other hand we likely will have users depend on it at some point.